docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+
+           if ($value->has('address')) {
+               $person->address = $this->addressCodec->decode($value->get('address'));
+           }
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           $data = [
+               '_id' => $value->id,
+               'name' => $value->name,
+           ];
+
+           if ($value->address) {
+               $data['address'] = $this->addressCodec->encode($value->address);
+           }
+
+           return MongoDB\BSON\Document::fromPHP($data);
+       }
+   }
+
+Codec Libraries
+---------------
+
+If you have a number of codecs that you want to use in multiple places, you can create a codec library. A codec library
+contains a list of codecs and checks each codec if it supports a value. If it does, it will use that codec to encode or
+decode the given value. The following code snippet changes the ``PersonCodec`` to use a codec library instead of a
+hard-coded ``DateTimeCodec``:
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       public function __construct(
+           private readonly MongoDB\Codec\CodecLibrary $codecLibrary,
+       ) {}
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->codecLibrary->decode($value->get('createdAt'));
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+               'createdAt' => $this->codecLibrary->encode($value->createdAt),
+           ]);
+       }
+   }
+
+This way, you can configure a global codec library and reuse it in multiple places:
+
+.. code-block:: php
+
+   <?php
+
+    $codecLibrary = new MongoDB\Codec\CodecLibrary([
+         new DateTimeCodec(),
+    ]);
+
+    $personCodec = new PersonCodec($codecLibrary);
+
+.. note::
+
+   Be careful when adding ``DocumentCodec`` instances to a codec library. The codec library will only use the first
+   codec that supports a given value. While this will most likely work fine when encoding to BSON, it can lead to issues
+   when decoding from BSON as you'll have to inspect the BSON document to detect whether the codec supports it. For this
```

Comment:
> lead to issues when decoding from BSON as you'll have to inspect the BSON document to detect whether the codec supports it.

Can you elaborate on this? The `encode()` method turns a PHP value into BSON, so are you saying that it's unlikely for multiple encoders to support the same value because they'd typically reject BSON? If so, how is that different for `decode()` methods if they typically receive BSON and spit out a PHP object?

Reply:
`encode` won't have an issue as it works on a specific class instance. `decode` on the other hand works on a `Document` and will have to inspect all fields to decide whether it can decode the given data. I agree that the wording is a bit awkward.

How about this:
> [...] it can lead to issues when decoding from BSON as the codec will have to inspect all fields in the BSON document to detect whether the codec actually supports it.

This would also provide an opportunity to talk about discriminator fields (e.g. the `odm` field used by `Persistable` or the `discriminatorField` functionality in Doctrine ODM).

Reply:
That language SGTM.

Reply:
Forgot this comment - the section was removed. I'll keep this language in mind if we add something like this again.